File: app/src/lib/contentMetrics.js
/**
 * A content post's metrics, read the same way everywhere.
 *
 * `perf` is written by three hands with three sets of names for one number:
 * PostBridge's sync (`view_count`), plans migrated from MaxAura (`views`,
 * `avgWatchTime`), and a person typing in what PostBridge cannot supply. Every
 * read goes through `METRIC_ALIASES`, which mirrors
 * backend/service/content_metrics.py — JavaScript cannot import the Python, so
 * backend/tests/test_content_metrics.py parses this table and fails when the
 * two differ, order included. Change both in the same commit.
 */

// Canonical name → perf keys that may carry it, highest priority first. The
// canonical name is also the manual-entry field and the key a typed value is
// stored under.
export const METRIC_ALIASES = Object.freeze({
  views:           ["view_count", "views", "play_count"],
  likes:           ["like_count", "likes", "digg_count"],
  comments:        ["comment_count", "comments"],
  shares:          ["share_count", "shares"],
  saves:           ["saves", "save_count", "collect_count"],
  reach:           ["reach"],
  avg_watch_time:  ["avg_watch_time", "avgWatchTime"],
  completion_rate: ["completion_rate", "completionRate"],
});

// What PostBridge's analytics carry, which the same test checks against its
// schema. On a post PostBridge published these sync on their own, so the form
// shows them rather than asking for them.
export const SYNCED_METRICS = Object.freeze(["views", "likes", "comments", "shares"]);

export const METRIC_NAMES = Object.freeze(Object.keys(METRIC_ALIASES));

function numberOrNull(v) {
  return typeof v === "number" && Number.isFinite(v) ? v : null;
}

/** One metric through its alias chain, or null. */
export function metricValue(perf, name) {
  for (const key of METRIC_ALIASES[name] || []) {
    const v = numberOrNull(perf?.[key]);
    if (v != null) return v;
  }
  return null;
}

/** Every metric by canonical name: `{ views, likes, …, completion_rate }`. */
export function metricsOf(perf) {
  return Object.fromEntries(METRIC_NAMES.map((name) => [name, metricValue(perf, name)]));
}

/** PostBridge published it, so its counts sync. */
export function isSyncedPost(post) {
  return Boolean(post?.post_bridge_post_id);
}

/** The metrics a person may type in on this post. */
export function editableMetrics(post) {
  return isSyncedPost(post)
    ? METRIC_NAMES.filter((name) => !SYNCED_METRICS.includes(name))
    : METRIC_NAMES;
}

/** The later of the last sync and the last hand edit, or null. */
export function lastUpdatedAt(perf) {
  const times = [perf?.last_synced_at, perf?.manual_updated_at]
    .map((v) => (v ? new Date(v).getTime() : NaN))
    .filter(Number.isFinite);
  return times.length ? new Date(Math.max(...times)).toISOString() : null;
}

/** The form's starting values: each metric as the string an input holds. */
export function metricInputs(perf, names) {
  return Object.fromEntries(
    names.map((name) => {
      const v = metricValue(perf, name);
      return [name, v == null ? "" : String(v)];
    }),
  );
}

/**
 * What the form changed, as the body POST /metrics takes: a number for a new
 * or corrected value, null for one the person cleared. A field left as it was
 * is not sent: sending it would mark a synced or migrated number as typed-in,
 * and from then on no sync could refresh it.
 */
export function metricChanges(inputs, perf, names) {
  const out = {};
  for (const name of names) {
    const raw = String(inputs?.[name] ?? "").trim();
    const current = metricValue(perf, name);
    if (raw === "") {
      if (current != null) out[name] = null;
      continue;
    }
    const n = Number(raw);
    if (Number.isFinite(n) && n !== current) out[name] = n;
  }
  return out;
}

File: backend/service/content_metrics.py
"""The one contract for reading and writing a content post's ``perf``.

``perf`` is a JSON column written by three hands, each with its own names for
the same number: PostBridge's sync (``view_count``, ``like_count`` …), plans
migrated from MaxAura (``views``, ``avgWatchTime``, ``completionRate`` …), and a
person typing in what PostBridge cannot supply. Reading a metric therefore means
trying its aliases in order, and ``METRIC_ALIASES`` is the only place that order
is written down. ``app/src/lib/contentMetrics.js`` carries the same table for
the browser (JavaScript cannot import this file); ``tests/test_content_metrics.py``
parses it and fails when the two disagree, because an alias one side knows and
the other does not is a number that shows in one place and not the other.

Writing has one rule: **a number a person entered is never overwritten by a
sync.** ``merge_manual_metrics`` records which metrics were typed in under
``manual_keys``; ``merge_synced_metrics`` skips every alias of those.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

# Canonical name → the perf keys that may carry it, highest priority first. The
# canonical name is also the manual-entry API field and the key a typed value is
# stored under, so it is always one of its own aliases. For the four counts
# PostBridge syncs its own key leads, because the sync is the freshest writer of
# those; for the rest the canonical key leads, because a person is.
METRIC_ALIASES: dict[str, tuple[str, ...]] = {
    "views":           ("view_count", "views", "play_count"),
    "likes":           ("like_count", "likes", "digg_count"),
    "comments":        ("comment_count", "comments"),
    "shares":          ("share_count", "shares"),
    "saves":           ("saves", "save_count", "collect_count"),
    "reach":           ("reach",),
    "avg_watch_time":  ("avg_watch_time", "avgWatchTime"),     # seconds
    "completion_rate": ("completion_rate", "completionRate"),  # percent, 0–100
}

MANUAL_KEYS = "manual_keys"              # the metrics a person entered, by canonical name
MANUAL_UPDATED_AT = "manual_updated_at"
LAST_SYNCED_AT = "last_synced_at"

_METRIC_OF_KEY: dict[str, str] = {
    key: name for name, keys in METRIC_ALIASES.items() for key in keys
}

# Never taken from a sync payload: "id" is PostBridge's analytics row, not a
# metric, and the bookkeeping keys are this module's to write, not a vendor's.
_NOT_SYNCED = frozenset({"id", MANUAL_KEYS, MANUAL_UPDATED_AT})


def _number(value: Any) -> int | float | None:
    # bool is an int to Python; a stray True is not one view.
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return value
    return None


def metric_value(perf: Mapping[str, Any] | None, name: str) -> int | float | None:
    """A metric through its alias chain, or None when no alias holds a number."""
    for key in METRIC_ALIASES[name]:
        value = _number((perf or {}).get(key))
        if value is not None:
            return value
    return None


def manual_metrics(perf: Mapping[str, Any] | None) -> set[str]:
    """The metrics a person entered on this post."""
    return set((perf or {}).get(MANUAL_KEYS) or [])


def save_rate(perf: Mapping[str, Any] | None) -> float | None:
    """Saves per view as a fraction, derived from the current numbers.

    Derived rather than read, because nothing writes a rate: PostBridge sends
    views and a person types saves, so a stored ``save_rate`` exists only on
    migrated rows and is the fallback for them.
    """
    views = metric_value(perf, "views")
    saves = metric_value(perf, "saves")
    if views and saves is not None:
        return saves / views
    return _number((perf or {}).get("save_rate"))


def merge_manual_metrics(
    perf: Mapping[str, Any] | None,
    entered: Mapping[str, int | float | None],
    *,
    at: str,
) -> dict[str, Any]:
    """Merge hand-entered metrics into a copy of ``perf`` and mark them manual.

    A value replaces the metric, not just its canonical key: every other alias
    is dropped, or a stale ``view_count`` earlier in the chain would still be
    what every reader sees. ``None`` withdraws a manual value, and unmarks it so
    the next sync owns the metric again.
    """
    unknown = set(entered) - set(METRIC_ALIASES)
    if unknown:
        raise ValueError(f"not a content metric: {sorted(unknown)}")
    merged = dict(perf or {})
    manual = manual_metrics(merged)
    for name, value in entered.items():
        for key in METRIC_ALIASES[name]:
            merged.pop(key, None)
        if value is None:
            manual.discard(name)
        else:
            merged[name] = value
            manual.add(name)
    merged[MANUAL_KEYS] = sorted(manual)
    merged[MANUAL_UPDATED_AT] = at
    return merged


def merge_synced_metrics(
    perf: Mapping[str, Any] | None,
    synced: Mapping[str, Any],
    *,
    synced_at: str | None,
) -> dict[str, Any]:
    """Merge a PostBridge analytics payload into a copy of ``perf``.

    Keys that alias a metric a person entered are skipped, so a sync can add
    and refresh numbers but never replace one somebody typed. Anything else in
    the payload (share URL, cover, platform ids) is written through as before.
    """
    merged = dict(perf or {})
    manual = manual_metrics(merged)
    for key, value in synced.items():
        if value is None or key in _NOT_SYNCED:
            continue
        if _METRIC_OF_KEY.get(key) in manual:
            continue
        merged[key] = value
    merged[LAST_SYNCED_AT] = synced_at
    return merged
